File: code/ch17/17.4.1.canMeasureWater.py
class Solution:
    def canMeasureWater(self, x: int, y: int, z: int) -> bool:
        if x + y < z:
            return False
        queue = [(0, 0)]
        seen = set((0, 0))

        while len(queue) > 0:
            a, b = queue.pop(0)
            if a == z or b == z or a + b == z:
                return True
            states = set()

            states.add((x, b))
            states.add((a, y))
            states.add((0, b))
            states.add((a, 0))
            states.add((min(x, b + a), 0 if b < x - a else b - (x - a)))
            states.add((0 if a + b < y else a - (y - b), min(b + a, y)))
            for state in states:
                if state in seen:
                    continue
                queue.append(state)
                seen.add(state)
        return False

File: code/ch17/17.4.1.canMeasureWater.cpp
#include <utility>
#include <set>
#include <queue>
#include <algorithm>
using namespace std;

class Solution {
public:
  bool canMeasureWater(int x, int y, int z) {
    if (x + y < z) return false;
    queue<pair<int, int> > q;
    q.push(make_pair(0, 0));
    set<pair<int, int> > seen;
    seen.insert(make_pair(0, 0));
    while (q.size()) {
      pair<int, int> buf = q.front();
      q.pop();
      int a = buf.first, b = buf.second;
      if (a == z || b == z || a + b == z) return true;
      set<pair<int, int> > states;
      states.insert(make_pair(x, b));
      states.insert(make_pair(a, y));
      states.insert(make_pair(0, b));
      states.insert(make_pair(a, 0));
      int val = ((b < x - a) ? 0 : (b - (x - a)));
      states.insert(make_pair(min(x, b + a), val));
      val = ((a + b < y) ? 0 : (a - (y - b)));
      states.insert(make_pair(val, min(b + a, y)));
      for (set<pair<int, int> >::iterator it = \
          states.begin(); it != states.end(); it++) {
        if (seen.count(*it)) continue;
        q.push(*it);
        seen.insert(*it);
      }
    }
    return false;
  }
};
